### PR TITLE
GFORMS-2533: Add Conditional Validation to Schema

### PR DESCRIPTION
**Pull request by Aashvin**

### CODE REVIEW

File: conf/formTemplateSchema.json [hunk 1]
```diff
@@ -729,7 +729,215 @@
             "$ref": "#/$defs/fields"
           }
         }
-      }
+      },
+      "allOf": [
+        {
+          "anyOf": [
+            {
+              "not": {"required": ["infoText"]}
```
Comment (josef-vlach) [inline conf/formTemplateSchema.json:737]: Would it be possible to format this? (applies to whole json)

My formatter do this for example:
```
              "not": {
                "required": [
                  "infoText"
                ]
              }
```

**Review by josef-vlach — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/gform/formtemplate/JsonSchemeValidator.scala [hunk 2]
```diff
@@ -32,20 +34,67 @@ object JsonSchemeValidator {
...
+                }.toList
+
+              errors.map(_.getMessage).toList.slice(0, indexOfAnyOf) ++ formattedConditionalValidationErrors
+            } else {
+              errors.map(_.getMessage).toList
+            }
+          }
+
+          SchemaValidationException(
+            NonEmptyList(conditionalValidationErrorMessages.head, conditionalValidationErrorMessages.tail)
```
Comment (lxol) [inline app/uk/gov/hmrc/gform/formtemplate/JsonSchemeValidator.scala:95]: is that possible that `conditionalValidationErrorMessages is empty?
Also you can use NonEmptyList.fromList method instead here or its unsafe version.

Comment (lxol) [reply app/uk/gov/hmrc/gform/formtemplate/JsonSchemeValidator.scala:95]: also `errors` on line 65 is NonEmptyList and the final result is NonEmptyList. May be we should keep it as `Nel` and never convert it to list, to ensure non-emptiness?

Comment (lxol) [reply app/uk/gov/hmrc/gform/formtemplate/JsonSchemeValidator.scala:95]: oops, just saw @josef-vlach comments about the same...

File: test/uk/gov/hmrc/gform/formtemplate/JsonSchemeValidatorSpec.scala [hunk 3]
```diff
@@ -0,0 +1,719 @@
...
+  test(
+    "validateJson rejects the form gracefully when the property infoType is used and the type property is not [info]"
+  ) {
+    val jsonTemplate = constructTestOneSectionJsonTemplate(
+      propertyDependencies = Map("type" -> "text"),
+      strings = Map("infoType" -> "noformat"),
+      ints = Map(),
+      arraysOfStrings = Map(),
+      arraysOfObjects = Map(),
+      objects = Map()
```
Comment (josef-vlach) [inline test/uk/gov/hmrc/gform/formtemplate/JsonSchemeValidatorSpec.scala:100]: Can we replace all these parameters with actual json under test? Like:
```
json"""
  {
    "infoType": "noformat",
    "type": "text",
    "id": "testId",
    "label": "test label"
  }"""
```
That would make the tests really nice to read.

**Review by josef-vlach — CHANGES_REQUESTED** [review]

**Review by josef-vlach — APPROVED** [review]
Please use Squash and merge 🙏

**Pull request merged**